Source/Core/Core/HW/SI_DeviceGBA.cpp

```diff
@@ -8,18 +8,75 @@
 #include "Common/StdMakeUnique.h"
 #include "Common/Thread.h"
 #include "Common/Logging/Log.h"
+#include "Core/CoreTiming.h"
 #include "Core/HW/SI_Device.h"
 #include "Core/HW/SI_DeviceGBA.h"
+#include "Core/HW/SystemTimers.h"
+#include "Core/HW/VideoInterface.h"
 
 #include "SFML/Network.hpp"
 
 static std::thread connectionThread;
+static std::thread clockThread;
 static std::queue<std::unique_ptr<sf::TcpSocket>> waiting_socks;
+static std::queue<std::unique_ptr<sf::TcpSocket>> waiting_clocks;
 static std::mutex cs_gba;
-namespace { volatile bool server_running; }
+static std::mutex cs_gba_clk;
+static u8 num_connected;
+
+namespace { volatile bool server_running; volatile bool clock_server_running; }
+
+enum EJoybusCmds
+{
+	CMD_RESET  = 0xff,
+	CMD_STATUS = 0x00,
+	CMD_READ   = 0x14,
+	CMD_WRITE  = 0x15
+};
+
+const u64 BITS_PER_SECOND = 115200;
+const u64 BYTES_PER_SECOND = BITS_PER_SECOND / 8;
+const u64 NUM_PORTS = 4;
+
+u8 GetNumConnected()
+{
+	int num_ports_connected = num_connected;
+	if (num_ports_connected == 0)
+		num_ports_connected = 1;
+
+	return num_ports_connected;
+}
 
 // --- GameBoy Advance "Link Cable" ---
 
+int GetTransferTime(u8 cmd)
+{
+	u64 bytes_transferred = 0;
+
+	switch (cmd)
+	{
+	case CMD_RESET:
+	case CMD_STATUS:
+	{
+		bytes_transferred = 4;
+		break;
+	}
+	case CMD_READ:
+	{
+		bytes_transferred = 6;
+		break;
+	}
+	case CMD_WRITE:
+	{
+		bytes_transferred = 1;
+		break;
+	}
+	default:
+		bytes_transferred = 1;
```

Comment:
Might want to add a break here to avoid getting this flagged on higher warning levels.